volatility3/framework/plugins/windows/pe_symbols.py

```diff
@@ -0,0 +1,777 @@
+        None,
+        None,
+    ]:
+        """
+        Yields each set of vads for a process that have a file mapped, along with the process itself and its layer
+        """
+        procs = pslist.PsList.list_processes(
+            context=context,
+            layer_name=layer_name,
+            symbol_table=symbol_table_name,
+        )
+
+        for proc in procs:
+            try:
+                proc_layer_name = proc.add_process_layer()
+            except exceptions.InvalidAddressException:
+                continue
+
+            vads = PESymbols.get_proc_vads_with_file_paths(proc)
+
+            yield proc, proc_layer_name, vads
+
+    @staticmethod
+    def get_process_modules(
+        context: interfaces.context.ContextInterface,
+        layer_name: str,
+        symbol_table: str,
+        filter_modules: Optional[filter_modules_type],
+    ) -> Dict[str, List[Tuple[str, int, int]]]:
+        """
+        Walks the process list and each process' VAD to determine the base address and size of wanted modules
+        """
+        proc_modules: Dict[str, List[Tuple[str, int, int]]] = {}
+
+        if filter_modules:
+            # create a tuple of module names for use with `endswith`
+            filter_modules_check = tuple([key.lower() for key in filter_modules.keys()])
+        else:
+            filter_modules_check = None
+
+        for _, proc_layer_name, vads in PESymbols.get_all_vads_with_file_paths(
+            context, layer_name, symbol_table
+        ):
+            for vad_start, vad_size, filepath in vads:
+                filename = PESymbols.filename_for_path(filepath)
+
+                if filter_modules_check and not filename.endswith(filter_modules_check):
+                    continue
+
+                # track each module along with the process layer and range to find it
+                if filename not in proc_modules:
+                    proc_modules[filename] = []
+
+                proc_modules[filename].append((proc_layer_name, vad_start, vad_size))
+
+        return proc_modules
+
+    def _generator(self) -> Generator[Tuple[int, Tuple[str, str, int]], None, None]:
+        kernel = self.context.modules[self.config["kernel"]]
+
+        if self.config["symbol"]:
+            filter_module = {
+                self.config["module"].lower(): {
+                    PESymbols.wanted_names: [self.config["symbol"]]
+                }
+            }
+
+        elif self.config["address"]:
+            filter_module = {
+                self.config["module"].lower(): {
+                    PESymbols.wanted_addresses: [self.config["address"]]
+                }
+            }
+
+        else:
+            vollog.error("--address or --symbol must be specified")
+            return
+
+        if self.config["source"] == "kernel":
+            module_resolver = self.get_kernel_modules
+        else:
+            module_resolver = self.get_process_modules
+
+        collected_modules = module_resolver(
```

Comment:
If this takes a while, it might be worth debugging lines or even calling the `progress_callback` to say what's going on while the setup's being done?

Reply:
I don't consider it slow enough to warrant that, its more that many of these malware checking plugins, like debug registers, will only trigger on a very small subset of samples - so this saves that work only for when its needed. So if people batch vol3 across 50 or 100+ samples at once, there is no need to gather this info for all of those when its likely none of them may even trigger output.